lib/signal-handler.c

```diff
@@ -23,74 +23,85 @@
  */
 
 #include "syslog-ng.h"
-#include "children.h"
 
 #include <signal.h>
 
-#if SYSLOG_NG_HAVE_DLFCN_H
-
-#include <dlfcn.h>
+#define SIGNAL_HANDLER_ARRAY_SIZE 128
 
-static const struct sigaction *sgchld_handler;
+static const struct sigaction *external_sigactions[SIGNAL_HANDLER_ARRAY_SIZE] = { NULL };
+static gboolean internal_sigaction_registered[SIGNAL_HANDLER_ARRAY_SIZE] = { FALSE };
 
 void
-trigger_sigchld_handler_chain(int signum)
+signal_handler_exec_external_handler(gint signum)
 {
-  if (sgchld_handler && sgchld_handler->sa_handler)
-    {
-      sgchld_handler->sa_handler(signum);
-    }
+  g_assert(signum < SIGNAL_HANDLER_ARRAY_SIZE);
+
+  const struct sigaction *external_sigaction = external_sigactions[signum];
+
+  if (!external_sigaction || !external_sigaction->sa_handler)
+    return;
+
+  external_sigaction->sa_handler(signum);
 }
 
+#if SYSLOG_NG_HAVE_DLFCN_H
+
+#include <dlfcn.h>
+
 static int
-call_original_sigaction(int signum, const struct sigaction *act, struct sigaction *oldact)
+_original_sigaction(int signum, const struct sigaction *act, struct sigaction *oldact)
 {
   static int (*real_sa)(int, const struct sigaction *, struct sigaction *);
 
   if (real_sa == NULL)
-    {
-      real_sa = dlsym(RTLD_NEXT, "sigaction");
-    }
+    real_sa = dlsym(RTLD_NEXT, "sigaction");
+
   return real_sa(signum, act, oldact);
 }
 
 static gboolean
-_save_handler(const struct sigaction *act)
+_need_to_save_external_sigaction_handler(gint signum)
 {
-  static gboolean is_first_handler = TRUE;
-  if (is_first_handler)
+  /* We need to save external sigaction handlers for signums we internally set a handler to. See lib/mainloop.c */
+  switch (signum)
     {
-      is_first_handler = FALSE;
+    case SIGCHLD:
+      g_assert(signum < SIGNAL_HANDLER_ARRAY_SIZE);
+      return TRUE;
+    default:
       return FALSE;
     }
+}
 
-  sgchld_handler = act;
-
-  child_manager_register_external_sigchld_handler(&trigger_sigchld_handler_chain);
+static void
+_save_external_sigaction_handler(gint signum, const struct sigaction *external_sigaction)
+{
+  if (external_sigaction && external_sigaction->sa_handler == SIG_DFL)
+    return;
 
-  return TRUE;
+  external_sigactions[signum] = external_sigaction;
 }
 
 /* This should be as defined in the <signal.h> */
 int
 sigaction(int signum, const struct sigaction *act, struct sigaction *oldact)
 {
+  if (!_need_to_save_external_sigaction_handler(signum))
+    return _original_sigaction(signum, act, oldact);
 
-  if (signum == SIGCHLD)
+  /* Internal sigactions are always the first one to arrive to this function. */
```

Comment:
Generally, I am fine with this patchset if it works, but I think we could be safer. Marking this comment/request as optional:

This patchset expects that `setup_signals` is the first one that registers the signal hanlders. This is not true, if a shared library modifies signal handlers in constructor functions.

According to the man page of `sigaction`, the old signal handler needs to be stored in `oldact`. For the first two calls of sigaction, this will happen through `_original_sigaction` for the same signal, but not for later calls. If we do not want to deal with this right now, it might worth adding a NULL assert for oldact.

This latter also breaks the scenario if two modules want to edit the same signal handler (additionally to `setup_signals`). The subsequent signal handlers just overwrite each other, and the original action is not propagated back by the output parameter. So the caller does not know about the original action, and cannot save and call it.

I was thinking the best would be if we did not need to override `sigaction` at all. If the signals weren't registered through ivykis, then we could just store `oldact` and call it in our signal handler. But it is registered by `iv_signal_register` currently.

Reading the ivykis manual, i found:

```
     It  is  permitted  to  register  multiple  struct  iv_signal
     objects for the same signal number.  If all such objects are
     registered with  IV_SIGNAL_FLAG_EXCLUSIVE in  ->flags unset,
     then all  objects will have their  callback functions called
     upon arrival of the signal, in no guaranteed order.
```

I am just guessing, but maybe if if we would not set this flag in `_register_signal_handler`, at least for the underlying signals, then maybe this sigaction implementation would not be necessary.

Another alternative: a bit hacky, but you can still imitate the sigaction behavior with `iv_signal_register`. You can get `oldact` with two subsequent sigaction calls: swapping out the handler with null, then back with the original value. Once you have oldact, you can properly implement the chaining without hacking `sigaction`.

Reply:
We've discussed that in the case of the Oracle instant client, overriding `sigaction` is necessary due to a possible bug in the client library.
We should copy `const struct sigaction *act`, because it is usually stored on the stack, and should return the appropriate `oldact` in all cases.

Reply:
@furiel Thanks for the comments.
@MrAnno Thanks, done all of that.